sqlmesh/core/engine_adapter/base.py

```diff
@@ -291,26 +293,63 @@ def drop_table(self, table_name: str, exists: bool = True) -> None:
     def alter_table(
         self,
         table_name: TableName,
-        added_columns: t.Dict[str, str],
-        dropped_columns: t.Sequence[str],
+        operations: t.List[SchemaDelta],
     ) -> None:
-        with self.transaction(TransactionType.DDL):
-            alter_table = exp.AlterTable(this=exp.to_table(table_name))
-
-            for column_name in dropped_columns:
-                drop_column = exp.Drop(this=exp.column(column_name), kind="COLUMN")
-                alter_table.set("actions", [drop_column])
-
-                self.execute(alter_table)
-
-            for column_name, column_type in added_columns.items():
-                add_column = exp.ColumnDef(
-                    this=exp.to_identifier(column_name),
-                    kind=parse_one(column_type, into=exp.DataType),  # type: ignore
+        def get_add_statement(operation: SchemaDelta) -> exp.AlterTable:
```

Comment:
each struct operation must be in it's own alter table? you've changed alter table to do multiple statements

Reply:
So the important thing is that in some cases we need the prior statement to execute before the next one can. For example if I add column `B` between `A` and `D` and then I want to add column `C` after `B` then either the engine needs to understand that dependency or we need to apply separately. Also we need to make sure we always resolve drops before adding for similar reasons. Therefore we could certainly batch up some of these operations but I decided to split them out in order to ensure things are executed accurately. If we get feedback about altering taking too long then we can look into batching these up but I'm betting this will not be an issue in practice.

Reply:
ok that's pretty nuanced, add a comment explaining why you did this